pkg/cloud/openstack/clients/machineservice.go

```diff
@@ -630,6 +638,48 @@ func (is *InstanceService) GetInstance(resourceId string) (instance *Instance, e
 	return serverToInstance(server), err
 }
 
+func (InstanceService) buildSchedulerHints(config *openstackconfigv1.OpenstackProviderSpec, opts servers.CreateOptsBuilder) (servers.CreateOptsBuilder, error) {
+	if config.SchedulerHints == nil {
+		return opts, nil
+	}
+	hints := queryFixer{
+		SchedulerHints: schedulerhints.SchedulerHints{
+			Group:           config.SchedulerHints.Group,
+			DifferentHost:   config.SchedulerHints.DifferentHost,
+			SameHost:        config.SchedulerHints.SameHost,
+			TargetCell:      config.SchedulerHints.TargetCell,
+			BuildNearHostIP: config.SchedulerHints.BuildNearHostIP,
+		},
+		Query: config.SchedulerHints.Query,
+	}
+	klog.Infof("adding schedulerHints to: %+v", hints)
+	return schedulerhints.CreateOptsExt{
+		CreateOptsBuilder: opts,
+		SchedulerHints:    hints,
+	}, nil
+}
+
+type queryFixer struct {
+	schedulerhints.SchedulerHints
+
+	Query string
+}
+
+func (opts queryFixer) ToServerSchedulerHintsCreateMap() (map[string]interface{}, error) {
+	// remove potential Query parameter from SchedulerHints
+	// Might not work with any version of openstack
```

Comment:
I tested the gophercloud `schedulerhints.SchedulerHints` against OpenStack Pike, which refuses to accept the json form (generated by gophercloud), but does accept the string form.

Also the json format would be a bit more complicated to use, because we can't store `[]interface{}`. It would need to be a string, which needs to be `json.Marshal`-ed. I found it easier to reason about just passing the string.

Any recommendation for a better comment?

Reply:
"Might not work with all versions of OpenStack" ?

Reply:
yes, this comment is better